File: Editorial/42_Trapping_Rain_Water/CSharp/3_Using_stacks.cs
public class Solution {
    public int Trap(int[] height) {
        int ans = 0, current = 0;
        Stack<int> st = new Stack<int>();
        while (current < height.Length) {
            while (st.Count != 0 && height[current] > height[st.Peek()]) {
                int top = st.Peek();
                st.Pop();
                if (st.Count == 0)
                    break;
                int distance = current - st.Peek() - 1;
                int bounded_height =
                    Math.Min(height[current], height[st.Peek()]) - height[top];
                ans += distance * bounded_height;
            }

            st.Push(current++);
        }

        return ans;
    }
}

File: Editorial/42_Trapping_Rain_Water/JavaScript/3_Using_stacks.js
var trap = function (height) {
    var ans = 0;
    var current = 0;
    var st = [];
    while (current < height.length) {
        while (st.length !== 0 && height[current] > height[st[st.length - 1]]) {
            var top = st[st.length - 1];
            st.pop();
            if (st.length === 0) break;
            var distance = current - st[st.length - 1] - 1;
            var bounded_height =
                Math.min(height[current], height[st[st.length - 1]]) -
                height[top];
            ans += distance * bounded_height;
        }
        st.push(current++);
    }
    return ans;
};
